app/code/Magento/CatalogUrlRewrite/Observer/CategoryUrlPathAutogeneratorObserver.php

```diff
@@ -107,33 +111,70 @@ public function execute(\Magento\Framework\Event\Observer $observer)
      *
      * @param Category $category
      * @param string $urlKey
-     * @throws \Magento\Framework\Exception\LocalizedException
-     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     *
+     * @throws LocalizedException
+     * @throws LogicException
+     * @throws InvalidArgumentException
+     * @throws NoSuchEntityException
      */
     private function updateUrlKey($category, $urlKey)
     {
+        $this->validateUrlKey($category, $urlKey);
+        $category->setUrlKey($urlKey)
+            ->setUrlPath($this->categoryUrlPathGenerator->getUrlPath($category));
+        if (!$category->isObjectNew()) {
+            $category->getResource()->saveAttribute($category, 'url_path');
+            if ($category->dataHasChangedFor('url_path')) {
+                $this->updateUrlPathForChildren($category);
+            }
+        }
+    }
+
+    /**
+     * Validate URL key value
+     *
+     * @param Category $category
+     * @param string $urlKey
+     *
+     * @throws LocalizedException
+     * @throws LogicException
+     */
+    private function validateUrlKey($category, $urlKey)
```

Comment:
Could you add argument and return types for this method?
And the same for updateUrlKey method

Reply:
The changes has been added, thanks.